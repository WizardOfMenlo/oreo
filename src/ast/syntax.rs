//! Typed syntax that is backed by Node

use super::node_db::{NodeDb, NodeId};
use super::untyped::{Node, NodeType};

/// Macro for simple definition of a syntax node
macro_rules! syntax_node {
    ($id:ident) => {
        /// A syntax node for an $id
        #[derive(Debug, Clone, Copy, Hash, Eq, PartialEq)]
        pub struct $id(NodeId);

        impl $id {
            /// Instantiate the syntax node
            pub fn new(node: NodeId) -> Self {
                $id(node)
            }

            /// Get node id
            pub fn get_id(self) -> NodeId {
                self.0
            }

            /// Get the raw node
            pub fn get_node<'a, 'b>(self, db: &'b NodeDb<'a>) -> &'b Node<'a> {
                db.get_node(self.0).expect("Invalid syntax node")
            }

            /// Get children
            pub fn children<'a, 'b>(self, db: &'b NodeDb<'a>) -> &'b [NodeId] {
                db.get_flat_node(self.0)
                    .expect("Invalid syntax node")
                    .children()
            }
        }
    };
}

syntax_node!(Program);

impl Program {
    /// Get the program name
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// Program main body
    pub fn compound<'a>(self, db: &NodeDb<'a>) -> Compound {
        Compound(self.children(db)[1])
    }
}

syntax_node!(Compound);

impl Compound {
    /// Get the inner statements
    pub fn statements<'a>(self, db: &NodeDb<'a>) -> Vec<Statement> {
        self.children(db).iter().map(|n| Statement(*n)).collect()
    }
}

syntax_node!(Statement);

impl Statement {
    /// Convert the statement to a more specific one
    pub fn downcast<'a>(self, db: &NodeDb<'a>) -> StatementType {
        let n = self.0;
        match self.get_node(db).ty() {
            NodeType::Decl => StatementType::Decl(Decl(n)),
            NodeType::PrintStat => StatementType::PrintStat(PrintStat(n)),
            NodeType::While => StatementType::While(While(n)),
            NodeType::If => StatementType::If(If(n)),
            NodeType::Assign => StatementType::Assign(Assign(n)),
            NodeType::FunctionCall => StatementType::FunctionCall(FunctionCall(n)),
            NodeType::FunctionDecl => StatementType::FunctionDecl(FunctionDecl(n)),
            NodeType::Return => StatementType::Return(Return(n)),
            _ => panic!("Invalid node constructed"),
        }
    }
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
/// The possible types that a statement can be
pub enum StatementType {
    Decl(Decl),
    PrintStat(PrintStat),
    While(While),
    If(If),
    Assign(Assign),
    FunctionCall(FunctionCall),
    FunctionDecl(FunctionDecl),
    Return(Return),
}

syntax_node!(Decl);

impl Decl {
    /// The identifier being declared
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// The expression to assign (if any)
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Option<Expr> {
        self.children(db).get(1).map(|n| Expr(*n))
    }
}

syntax_node!(PrintStat);

impl PrintStat {
    /// Convert to the more specific print typ
    pub fn downcast<'a>(self, db: &NodeDb<'a>) -> PrintTypes {
        let node = self.children(db)[0];
        match self.get_node(db).ty() {
            NodeType::Print => PrintTypes::Print(Print(node)),
            NodeType::Println => PrintTypes::Println(Println(node)),
            NodeType::Get => PrintTypes::Get(Get(node)),
            _ => panic!("Invalid node constructed"),
        }
    }
}

/// Possible print types
#[derive(Debug, Clone)]
#[allow(missing_docs)]
pub enum PrintTypes {
    Print(Print),
    Println(Println),
    Get(Get),
}

syntax_node!(Print);

impl Print {
    /// The expression to print
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }
}

syntax_node!(Println);

impl Println {
    /// The expression to print
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }
}

syntax_node!(Get);

impl Get {
    /// The identifier to set
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }
}

syntax_node!(While);

impl While {
    /// The condition to evaluate
    pub fn condition<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }

    /// The body to be executed
    pub fn compound<'a>(self, db: &NodeDb<'a>) -> Compound {
        Compound(self.children(db)[1])
    }
}

syntax_node!(If);

impl If {
    /// The condition to evaluate
    pub fn condition<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }

    /// If true
    pub fn if_branch<'a>(self, db: &NodeDb<'a>) -> Compound {
        Compound(self.children(db)[1])
    }

    /// If false (optional)
    pub fn else_branch<'a>(self, db: &NodeDb<'a>) -> Option<Compound> {
        self.children(db).get(2).map(|n| Compound(*n))
    }
}

syntax_node!(Assign);

impl Assign {
    /// The identifier to set
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// The expression to set it to
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[1])
    }
}

syntax_node!(FunctionDecl);

impl FunctionDecl {
    /// Function to declare
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// The arguments
    pub fn args<'a>(self, db: &NodeDb<'a>) -> Vec<FunctionDeclArgs> {
        let children = self.children(db);
        children[1..children.len() - 1]
            .iter()
            .map(|n| FunctionDeclArgs(*n))
            .collect()
    }

    /// The body to be executed
    pub fn compound<'a>(self, db: &NodeDb<'a>) -> Compound {
        let children = self.children(db);
        Compound(children[children.len() - 1])
    }
}

syntax_node!(FunctionDeclArgs);

impl FunctionDeclArgs {
    /// The arg id
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }
}

syntax_node!(FunctionCallArgs);

impl FunctionCallArgs {
    /// The expression in the argument
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }
}

syntax_node!(Return);

impl Return {
    /// The expression to return
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }
}

syntax_node!(FunctionCall);

impl FunctionCall {
    /// The function id to call
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// The arguments of the call
    pub fn args<'a>(self, db: &NodeDb<'a>) -> Vec<FunctionCallArgs> {
        self.children(db)[1..]
            .iter()
            .map(|n| FunctionCallArgs(*n))
            .collect()
    }
}

syntax_node!(Expr);

impl Expr {
    /// The head of the expression
    pub fn head<'a>(self, db: &NodeDb<'a>) -> Term {
        Term(self.children(db)[0])
    }

    /// Tail of the expression
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<ExprPrime> {
        self.children(db).get(1).map(|e| ExprPrime(*e))
    }
}

syntax_node!(ExprPrime);

impl ExprPrime {
    /// The operation
    pub fn op<'a>(self, db: &NodeDb<'a>) -> BooleanOp {
        self.get_node(db).ty().clone().into()
    }

    /// The right operand
    pub fn rhs<'a>(self, db: &NodeDb<'a>) -> Term {
        Term(self.children(db)[0])
    }

    /// Possible tails
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<ExprPrime> {
        self.children(db).get(1).map(|e| ExprPrime(*e))
    }
}

syntax_node!(Term);

impl Term {
    /// Head of the term
    pub fn head<'a>(self, db: &NodeDb<'a>) -> Factor {
        Factor(self.children(db)[0])
    }

    /// Tail of the term
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<TermPrime> {
        self.children(db).get(1).map(|n| TermPrime(*n))
    }
}

syntax_node!(TermPrime);

impl TermPrime {
    /// The operation
    pub fn op<'a>(self, db: &NodeDb<'a>) -> RelationalOp {
        self.get_node(db).ty().clone().into()
    }

    /// Rhs
    pub fn rhs<'a>(self, db: &NodeDb<'a>) -> Factor {
        Factor(self.children(db)[0])
    }

    /// Optional tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<TermPrime> {
        self.children(db).get(1).map(|n| TermPrime(*n))
    }
}

syntax_node!(Factor);

impl Factor {
    /// The head of a factor
    pub fn head<'a>(self, db: &NodeDb<'a>) -> Product {
        Product(self.children(db)[0])
    }

    /// Optional tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<FactorPrime> {
        self.children(db).get(1).map(|n| FactorPrime(*n))
    }
}

syntax_node!(FactorPrime);

impl FactorPrime {
    /// The operation in the tail
    pub fn op<'a>(self, db: &NodeDb<'a>) -> AdditiveOp {
        self.get_node(db).ty().clone().into()
    }

    /// The rhs of the tail
    pub fn rhs<'a>(self, db: &NodeDb<'a>) -> Product {
        Product(self.children(db)[0])
    }

    /// Optional tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<FactorPrime> {
        self.children(db).get(1).map(|n| FactorPrime(*n))
    }
}

syntax_node!(Product);

impl Product {
    /// The head
    pub fn head<'a>(self, db: &NodeDb<'a>) -> Atom {
        Atom(self.children(db)[0])
    }

    /// Opt tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<ProductPrime> {
        self.children(db).get(1).map(|n| ProductPrime(*n))
    }
}

syntax_node!(ProductPrime);

impl ProductPrime {
    /// The operation
    pub fn op<'a>(self, db: &NodeDb<'a>) -> MultiplicativeOp {
        self.get_node(db).ty().clone().into()
    }

    /// The rhs of the product
    pub fn rhs<'a>(self, db: &NodeDb<'a>) -> Atom {
        Atom(self.children(db)[0])
    }

    /// Optional tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<ProductPrime> {
        self.children(db).get(1).map(|n| ProductPrime(*n))
    }
}

syntax_node!(Atom);

/// Possible types of an atom
pub enum AtomType {
    /// Literal, function call or bracketed expr
    Unit(Unit),

    /// The not of an atom
    Not(Atom),
}

impl Atom {
    /// Convert an atom to a more specific type
    pub fn downcast<'a>(self, db: &NodeDb<'a>) -> AtomType {
        let n = self.children(db)[0];
        match self.get_node(db).ty() {
            NodeType::Not => AtomType::Not(Atom(n)),
            NodeType::Unit => AtomType::Unit(Unit(n)),
            _ => panic!("Invalid atom type"),
        }
    }
}

syntax_node!(Unit);

impl Unit {
    /// Convert a unit into a more specific type
    pub fn downcast<'a>(self, db: &NodeDb<'a>) -> UnitType {
        match self.get_node(db).ty() {
            NodeType::Int(i) => UnitType::Int(*i),
            NodeType::Bool(b) => UnitType::Bool(*b),
            NodeType::Str => UnitType::Str(self.0),
            NodeType::Identifier => UnitType::Identifier(Identifier(self.children(db)[0])),
            NodeType::FunctionCall => UnitType::FunctionCall(FunctionCall(self.children(db)[0])),
            NodeType::BracketedExpr => UnitType::BracketedExpr(Expr(self.children(db)[0])),
            _ => panic!("Invalid atom type"),
        }
    }
}

/// Possible unit types
#[derive(Debug, Clone)]
#[allow(missing_docs)]
pub enum UnitType {
    Int(isize),
    Str(NodeId),
    Bool(bool),
    Identifier(Identifier),
    FunctionCall(FunctionCall),
    BracketedExpr(Expr),
}

syntax_node!(Identifier);

impl Identifier {
    /// Get the string id
    pub fn id<'a>(self, db: &NodeDb<'a>, text: &'a str) -> &'a str {
        use crate::range::Ranged;
        &text[self.get_node(db).range().clone()]
    }
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum AdditiveOp {
    Plus,
    Minus,
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum MultiplicativeOp {
    Times,
    Divide,
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum RelationalOp {
    LesserThan,
    GreaterThan,
    Equals,
    GreaterOrEquals,
    LesserOrEquals,
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum BooleanOp {
    And,
    Or,
}

/// Enum to convert any node into a typed one
#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum SyntaxNode {
    Program(Program),
    Compound(Compound),
    Decl(Decl),
    PrintStat(PrintStat),
    Print(Print),
    Println(Println),
    Get(Get),
    While(While),
    If(If),
    Assign(Assign),
    FunctionDecl(FunctionDecl),
    FunctionDeclArgs(FunctionDeclArgs),
    FunctionCall(FunctionCall),
    FunctionCallArgs(FunctionCallArgs),
    Return(Return),
    Expr(Expr),
    ExprPrime(ExprPrime),
    Term(Term),
    TermPrime(TermPrime),
    Factor(Factor),
    FactorPrime(FactorPrime),
    Product(Product),
    ProductPrime(ProductPrime),
    Atom(Atom),
    Unit(Unit),
    Identifier(Identifier),
}

impl SyntaxNode {
    /// Convert a id to a syntax node
    pub fn from<'a>(input: NodeId, db: &NodeDb<'a>) -> Self {
        match db.get_node(input).expect("Invalid id").ty() {
            NodeType::Program => SyntaxNode::Program(Program(input)),
            NodeType::Compound => SyntaxNode::Compound(Compound(input)),
            NodeType::Decl => SyntaxNode::Decl(Decl(input)),
            NodeType::PrintStat => SyntaxNode::PrintStat(PrintStat(input)),

            NodeType::While => SyntaxNode::While(While(input)),
            NodeType::If => SyntaxNode::If(If(input)),
            NodeType::Assign => SyntaxNode::Assign(Assign(input)),
            NodeType::FunctionCall => SyntaxNode::FunctionCall(FunctionCall(input)),
            NodeType::FunctionDecl => SyntaxNode::FunctionDecl(FunctionDecl(input)),
            NodeType::Return => SyntaxNode::Return(Return(input)),
            NodeType::Print => SyntaxNode::Print(Print(input)),
            NodeType::Println => SyntaxNode::Println(Println(input)),
            NodeType::Get => SyntaxNode::Get(Get(input)),
            NodeType::FunctionDeclArgs => SyntaxNode::FunctionDeclArgs(FunctionDeclArgs(input)),
            NodeType::FunctionCallArgs => SyntaxNode::FunctionCallArgs(FunctionCallArgs(input)),

            NodeType::Expr => SyntaxNode::Expr(Expr(input)),
            NodeType::And | NodeType::Or => SyntaxNode::ExprPrime(ExprPrime(input)),

            NodeType::Term => SyntaxNode::Term(Term(input)),
            NodeType::LesserThan
            | NodeType::GreaterThan
            | NodeType::Equals
            | NodeType::GreaterOrEquals
            | NodeType::LesserOrEquals => SyntaxNode::TermPrime(TermPrime(input)),

            NodeType::Factor => SyntaxNode::Factor(Factor(input)),
            NodeType::Plus | NodeType::Minus => SyntaxNode::FactorPrime(FactorPrime(input)),

            NodeType::Product => SyntaxNode::Product(Product(input)),
            NodeType::Times | NodeType::Divide => SyntaxNode::ProductPrime(ProductPrime(input)),

            NodeType::Unit | NodeType::Not => SyntaxNode::Atom(Atom(input)),

            NodeType::Int(_) | NodeType::Str | NodeType::Bool(_) | NodeType::BracketedExpr => {
                SyntaxNode::Unit(Unit(input))
            }

            NodeType::Identifier => SyntaxNode::Identifier(Identifier(input)),

            NodeType::Error(_) => panic!("Called on error node"),
        }
    }
}
