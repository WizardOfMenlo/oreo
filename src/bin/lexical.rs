use oreo::lexical::lexicalize;
use oreo::scanner::scan;
use structopt::StructOpt;

#[derive(Debug, StructOpt)]
struct Args {
    /// The input to parse
    #[structopt(short, long)]
    input: String,
}

fn main() {
    let opt = Args::from_args();
    println!("{:?}", lexicalize(scan(&opt.input)).collect::<Vec<_>>());
}
