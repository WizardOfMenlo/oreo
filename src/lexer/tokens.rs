//! Definition of all the tokens that we use

use super::error::LexicalError;
use serde::Serialize;

/// The main token type that we work on
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Token<'a> {
    /// Any of the keywords (apart from and and or)
    Keyword(Keyword),

    /// (, ), ;, ,
    Punctuation(Punctuation),

    /// Addition, mult, ...
    Operator(Operator),

    /// A constant
    Literal(Literal<'a>),

    /// An identifier
    Identifier(&'a str),

    /// A comment
    Comment(&'a str),

    /// Any kind of error we might encounter
    Error(LexicalError<'a>),
}

impl<'a> Token<'a> {
    /// Is the current token an error?
    pub fn is_error(&self) -> bool {
        match self {
            Token::Error(_) => true,
            _ => false,
        }
    }

    /// Is the current token a comment?
    pub fn is_comment(&self) -> bool {
        match self {
            Token::Comment(_) => true,
            _ => false,
        }
    }

    /// Is this token the same as an other (where Int(1) == Int(2))
    pub fn same_kind<'b>(&self, other: &Token<'b>) -> bool
    where
        'b: 'a,
    {
        // This should match every singleton
        if self == other {
            true
        } else {
            // Match literals
            match (self, other) {
                (Self::Literal(Literal::Integer(_)), Self::Literal(Literal::Integer(_))) => true,
                (Self::Literal(Literal::Boolean(_)), Self::Literal(Literal::Boolean(_))) => true,
                (Self::Literal(Literal::String(_)), Self::Literal(Literal::String(_))) => true,
                (Self::Identifier(_), Self::Identifier(_)) => true,
                (Self::Comment(_), Self::Comment(_)) => true,
                _ => false,
            }
        }
    }
}

/// The Keywords we have
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Keyword {
    Program,
    Begin,
    End,
    Var,
    Print,
    Println,
    Get,
    While,
    If,
    Then,
    Else,
    Procedure,
    Return,
}

/// Possible Punctuation marks
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Punctuation {
    Comma,
    Semicolon,
    BracketOpen,
    BracketClose,
}

/// All the operators (including boolean ones)
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Operator {
    Assignement,
    Plus,
    Minus,
    Times,
    Divide,
    LesserThan,
    GreaterThan,
    Equals,
    GreaterOrEquals,
    LesserOrEquals,
    Or,
    And,
    Not,
}

/// Constants
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Literal<'a> {
    Integer(isize),
    Boolean(bool),
    String(&'a str),
}
