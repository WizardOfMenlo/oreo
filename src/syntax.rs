pub struct Identifier<'a>(&'a str);

pub struct Program<'a> {
    id: Identifier<'a>,
    compund: Compound<'a>,
}

pub struct Compound<'a> {
    statements: Vec<Statement<'a>>,
}

pub enum Statement<'a> {
    Decl(Decl<'a>),
    Print(Print<'a>),
    While(While<'a>),
    If(If<'a>),
    Assign(Assign<'a>),
}

pub struct Decl<'a> {
    id: Identifier<'a>,
    expr: Option<Expr<'a>>,
}

pub enum Print<'a> {
    Print(Expr<'a>),
    Println(Expr<'a>),
    Get(Identifier<'a>),
}

pub struct While<'a> {
    condition: Bool<'a>,
    compound: Compound<'a>,
}

pub struct If<'a> {
    condition: Bool<'a>,
    if_branch: Compound<'a>,
    else_branch: Option<Compound<'a>>,
}

pub struct Assign<'a> {
    id: Identifier<'a>,
    expr: Expr<'a>,
}

pub enum Expr<'a> {
    Addition(Box<BinaryOperator<'a>>),
    Subtraction(Box<BinaryOperator<'a>>),
    Multiplication(Box<BinaryOperator<'a>>),
    Division(Box<BinaryOperator<'a>>),
    Bool(Box<Bool<'a>>),
    Expr(Box<Expr<'a>>),
    Literal(Literal<'a>),
    Identifier(Identifier<'a>)
}

pub enum Literal<'a> {
    Str(&'a str),
    Int(isize)
}

pub struct BinaryOperator<'a> {
    left: Expr<'a>,
    right: Expr<'a>,
}

pub enum Bool<'a> {
    LessThan(BooleanRelOperator<'a>),
    GreaterThan(BooleanRelOperator<'a>),
    LesserOrEquals(BooleanRelOperator<'a>),
    GreaterOrEquals(BooleanRelOperator<'a>),
    Equals(BooleanRelOperator<'a>),
    And(Box<BooleanBinaryOperator<'a>>),
    Or(Box<BooleanBinaryOperator<'a>>),
    Not(Box<Bool<'a>>),
    BooleanLiteral(bool),
}

pub struct BooleanRelOperator<'a> {
    first: Expr<'a>,
    second: Expr<'a>,
}

pub struct BooleanBinaryOperator<'a> {
    first: Bool<'a>,
    second: Bool<'a>,
}