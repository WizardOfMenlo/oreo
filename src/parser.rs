use crate::lexical::EnrichedToken;
use crate::parsing_utils::*;
use crate::syntax::*;
use crate::tokens::*;

pub(crate) type ExpToken = Token<'static>;
pub(crate) type TokenList = &'static [ExpToken];

pub enum SyntaxError<'a> {
    ExpectedOneOfButFoundEOF(TokenList),
    ExpectedButFoundEOF(ExpToken),
    ExpectedFound(ExpToken, EnrichedToken<'a>),
    ExpectedOneOfFound(TokenList, EnrichedToken<'a>),
    LogicalError,
}

pub(crate) const ID: ExpToken = Token::Identifier("");

pub fn program<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<Program<'a>, SyntaxError<'a>> {
    advance_expecting(it, Token::Keyword(Keyword::Program))?;
    let id = advance_expecting_identifier(it)?;
    let compound = compound(it)?;

    Ok(Program { id, compound })
}

fn compound<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<Compound<'a>, SyntaxError<'a>> {
    advance_expecting(it, Token::Keyword(Keyword::Begin))?;
    const END: ExpToken = Token::Keyword(Keyword::End);
    let mut statements = Vec::new();
    let mut it = it.peekable();
    loop {
        if let Some(tok) = it.peek() {
            let token = tok.token();
            if token.same_kind(&END) {
                break;
            }

            statements.push(statement(&mut it)?);
        } else {
            return Err(SyntaxError::ExpectedButFoundEOF(END));
        }
    }

    Ok(Compound { statements })
}

fn statement<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<Statement<'a>, SyntaxError<'a>> {
    const POSSIBLE_STATEMENT_STARTS: TokenList = &[
        Token::Keyword(Keyword::Var),
        Token::Keyword(Keyword::If),
        Token::Keyword(Keyword::Print),
        Token::Keyword(Keyword::Println),
        Token::Keyword(Keyword::Get),
        Token::Keyword(Keyword::While),
        ID,
    ];

    let matched = advance_expecting_one_of(it, POSSIBLE_STATEMENT_STARTS)?;
    let token = matched.take_token();
    match token {
        Token::Keyword(Keyword::Var) => Ok(Statement::Decl(decl(it)?)),
        Token::Keyword(Keyword::If) => Ok(Statement::If(p_if(it)?)),
        Token::Keyword(Keyword::Print) => Ok(Statement::Print(print(it, token)?)),
        Token::Keyword(Keyword::Println) => Ok(Statement::Print(print(it, token)?)),
        Token::Keyword(Keyword::Get) => Ok(Statement::Print(print(it, token)?)),
        Token::Keyword(Keyword::While) => Ok(Statement::While(p_while(it)?)),
        Token::Identifier(s) => Ok(Statement::Assign(assign(it, Identifier(s))?)),
        _ => Err(SyntaxError::LogicalError),
    }
}

fn decl<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<Decl<'a>, SyntaxError<'a>> {
    let id = advance_expecting_identifier(it)?;
    let mut it = it.peekable();
    let next = it.peek();

    match next.map(|s| s.token()) {
        Some(Token::Punctuation(Punctuation::Semicolon)) => Ok(Decl { id, expr: None }),
        _ => Ok(Decl { id, expr: Some(expr(&mut it)?)})
    }
}
 
fn p_if<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<If<'a>, SyntaxError<'a>> {
    advance_expecting(it, Token::Punctuation(Punctuation::BracketOpen))?;
    let condition = p_bool(it)?;
    advance_expecting(it, Token::Punctuation(Punctuation::BracketClose))?;
    advance_expecting(it, Token::Keyword(Keyword::Then))?;
    let if_branch = compound(it)?;
    let next = advance_expecting_one_of(it, &[Token::Keyword(Keyword::Else), Token::Punctuation(Punctuation::Semicolon)])?;

    let else_branch = match next.token() {
        Token::Keyword(Keyword::Else) => {
            let tmp = compound(it)?;
            advance_expecting(it, Token::Punctuation(Punctuation::Semicolon))?;
            Some(tmp)
        },
        _ => None
    };

    Ok(If {
        condition,
        if_branch,
        else_branch
    })
}

fn print<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
    token: Token<'a>
) -> Result<Print<'a>, SyntaxError<'a>> {
    match token {
        Token::Keyword(Keyword::Print) => Ok(Print::Print(expr(it)?)),
        Token::Keyword(Keyword::Println) => Ok(Print::Println(expr(it)?)),
        Token::Keyword(Keyword::Get) => Ok(Print::Get(advance_expecting_identifier(it)?)),
        // Should only be called with normal stuff
        _ => Err(SyntaxError::LogicalError)
    }
}

fn p_while<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<While<'a>, SyntaxError<'a>> {
    advance_expecting(it, Token::Punctuation(Punctuation::BracketOpen))?;
    let condition = p_bool(it)?;
    advance_expecting(it, Token::Punctuation(Punctuation::BracketClose))?;
    let compound = compound(it)?;
    advance_expecting(it, Token::Punctuation(Punctuation::Semicolon))?;

    Ok(While {
        condition, 
        compound
    })
}
 
fn assign<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
    id: Identifier<'a>
) -> Result<Assign<'a>, SyntaxError<'a>> {
    advance_expecting(it, Token::Operator(Operator::Assignement))?;
    let expr = expr(it)?;

    Ok(Assign {
        id,
        expr
    })
}
 
fn expr<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<Expr<'a>, SyntaxError<'a>> {
    Err(SyntaxError::LogicalError)
}

fn p_bool<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<Bool<'a>, SyntaxError<'a>> {
    Err(SyntaxError::LogicalError)
}