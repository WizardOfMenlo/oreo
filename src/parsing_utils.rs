use crate::lexical::EnrichedToken;
use crate::parser::{consts, ExpToken, SyntaxError, TokenList};
use crate::syntax::*;
use crate::tokens::*;

#[must_use]
pub fn advance_expecting<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
    exp: ExpToken,
) -> Result<EnrichedToken<'a>, SyntaxError<'a>> {
    let next = it.next();
    match next {
        Some(tok) => {
            let found_token = tok.token();
            if found_token.same_kind(&exp) {
                Ok(tok)
            } else {
                Err(SyntaxError::ExpectedFound(exp, tok))
            }
        }
        None => Err(SyntaxError::ExpectedButFoundEOF(exp)),
    }
}

#[must_use]
pub fn advance_expecting_one_of<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
    exp: TokenList,
) -> Result<EnrichedToken<'a>, SyntaxError<'a>> {
    let next = it.next();
    match next {
        Some(tok) => {
            let found_token = tok.token();
            if exp.iter().any(|t| found_token.same_kind(t)) {
                Ok(tok)
            } else {
                Err(SyntaxError::ExpectedOneOfFound(exp, tok))
            }
        }
        None => Err(SyntaxError::ExpectedOneOfButFoundEOF(exp)),
    }
}

#[must_use]
pub fn get_id(id: EnrichedToken) -> Result<Identifier, SyntaxError> {
    let backup = id.clone();
    match id.take_token() {
        Token::Identifier(s) => Ok(Identifier(s)),
        _ => Err(SyntaxError::ExpectedFound(consts::ID, backup)),
    }
}

#[must_use]
pub fn advance_expecting_identifier<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<Identifier<'a>, SyntaxError<'a>> {
    let next = it.next();
    match next {
        Some(tok) => get_id(tok),
        None => Err(SyntaxError::ExpectedButFoundEOF(consts::ID)),
    }
}
