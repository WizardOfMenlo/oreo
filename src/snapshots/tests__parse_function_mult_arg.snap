---
source: src/parser.rs
expression: parsed
---
Node {
    ty: Program,
    text_range: 0..73,
    children: [
        Node {
            ty: Identifier,
            text_range: 8..11,
            children: [],
        },
        Node {
            ty: Compound,
            text_range: 12..73,
            children: [
                Node {
                    ty: FunctionDecl,
                    text_range: 18..69,
                    children: [
                        Node {
                            ty: Identifier,
                            text_range: 28..31,
                            children: [],
                        },
                        Node {
                            ty: FunctionDeclArgs,
                            text_range: 32..38,
                            children: [
                                Node {
                                    ty: Identifier,
                                    text_range: 36..37,
                                    children: [],
                                },
                            ],
                        },
                        Node {
                            ty: FunctionDeclArgs,
                            text_range: 39..44,
                            children: [
                                Node {
                                    ty: Identifier,
                                    text_range: 43..44,
                                    children: [],
                                },
                            ],
                        },
                        Node {
                            ty: Compound,
                            text_range: 46..69,
                            children: [
                                Node {
                                    ty: Return,
                                    text_range: 52..65,
                                    children: [
                                        Node {
                                            ty: Expr,
                                            text_range: 59..64,
                                            children: [
                                                Node {
                                                    ty: Term,
                                                    text_range: 59..64,
                                                    children: [
                                                        Node {
                                                            ty: Factor,
                                                            text_range: 59..64,
                                                            children: [
                                                                Node {
                                                                    ty: Product,
                                                                    text_range: 59..60,
                                                                    children: [
                                                                        Node {
                                                                            ty: Unit,
                                                                            text_range: 59..60,
                                                                            children: [
                                                                                Node {
                                                                                    ty: Identifier,
                                                                                    text_range: 59..60,
                                                                                    children: [],
                                                                                },
                                                                            ],
                                                                        },
                                                                    ],
                                                                },
                                                                Node {
                                                                    ty: Plus,
                                                                    text_range: 61..64,
                                                                    children: [
                                                                        Node {
                                                                            ty: Product,
                                                                            text_range: 63..64,
                                                                            children: [
                                                                                Node {
                                                                                    ty: Unit,
                                                                                    text_range: 63..64,
                                                                                    children: [
                                                                                        Node {
                                                                                            ty: Identifier,
                                                                                            text_range: 63..64,
                                                                                            children: [],
                                                                                        },
                                                                                    ],
                                                                                },
                                                                            ],
                                                                        },
                                                                    ],
                                                                },
                                                            ],
                                                        },
                                                    ],
                                                },
                                            ],
                                        },
                                    ],
                                },
                            ],
                        },
                    ],
                },
            ],
        },
    ],
}
